libs/contrib/Data/List/Lazy.idr

```diff
@@ -10,6 +10,23 @@ data LazyList : Type -> Type where
   Nil : LazyList a
   (::) : (x : a) -> (xs : Lazy (LazyList a)) -> LazyList a
 
+--- Truly lazy functions ---
+
+public export
+(++) : LazyList a -> Lazy (LazyList a) -> LazyList a
+[]      ++ ys = ys
+(x::xs) ++ ys = x :: (xs ++ ys)
+
+public export
+foldrLazy : (func : elem -> Lazy acc -> acc) -> (init : acc) -> (input : LazyList elem) -> acc
```

Comment:
Shouldn't `init` be lazy too? If `func` throws away its second argument and `input`
is non-empty, there is no reason to evaluate `init`!

Reply:
I agree with you. I was solving a particular problem and wasn't thinking a little further.